main_test.go

```diff
@@ -17,53 +17,109 @@ limitations under the License.
 
 type mainTest struct {
-	name      string
-	input     string
-	golden    string
-	errString string
-}
-
-var mainTests = []mainTest{
-	{"numeric_literal", "100", "100", ""},
-	{"boolean_literal", "true", "true", ""},
-	{"simple_arith1", "3 + 3", "6", ""},
-	{"simple_arith2", "3 + 3 + 3", "9", ""},
-	{"simple_arith3", "(3 + 3) + (3 + 3)", "12", ""},
-	{"unicode", `"\u263A"`, `"☺"`, ""},
-	{"unicode2", `"\u263a"`, `"☺"`, ""},
-	{"escaped_single_quote", `"\\'"`, `"\\'"`, ""},
-	{"simple_arith_string", "\"aaa\" + \"bbb\"", "\"aaabbb\"", ""},
-	{"simple_arith_string2", "\"aaa\" + \"\"", "\"aaa\"", ""},
-	{"simple_arith_string3", "\"\" + \"bbb\"", "\"bbb\"", ""},
-	{"simple_arith_string_empty", "\"\" + \"\"", "\"\"", ""},
-	{"verbatim_string", `@"blah ☺"`, `"blah ☺"`, ""},
-	{"empty_array", "[]", "[ ]", ""},
-	{"array", "[1, 2, 1 + 2]", "[\n   1,\n   2,\n   3\n]", ""},
-	{"empty_object", "{}", "{ }", ""},
-	{"object", `{"x": 1+1}`, "{\n   \"x\": 2\n}", ""},
+	name   string
+	input  string
+	golden string
 }
 
 func TestMain(t *testing.T) {
-	for _, test := range mainTests {
-		vm := MakeVM()
-		output, err := vm.EvaluateSnippet(test.name, test.input)
-		var errString string
-		if err != nil {
-			errString = err.Error()
-		}
-		if errString != test.errString {
-			t.Errorf("%s: error result does not match. got\n\t%+v\nexpected\n\t%+v",
-				test.input, errString, test.errString)
-		}
-		if err == nil && output != test.golden {
-			t.Errorf("%s: got\n\t%+v\nexpected\n\t%+v", test.name, output, test.golden)
+	flag.Parse()
+	var mainTests []mainTest
+	match, err := filepath.Glob("testdata/*.input")
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, input := range match {
+		golden := input
+		name := input
+		if strings.HasSuffix(input, ".input") {
+			name = input[:len(input)-len(".input")]
+			golden = name + ".golden"
 		}
+		mainTests = append(mainTests, mainTest{name: name, input: input, golden: golden})
+	}
+	for _, test := range mainTests {
+		t.Run(test.name, func(t *testing.T) {
+			vm := MakeVM()
+			read := func(file string) []byte {
+				bytz, err := ioutil.ReadFile(file)
+				if err != nil {
+					t.Fatalf("reading file: %s: %v", file, err)
+				}
+				return bytz
+			}
+
+			input := read(test.input)
+			output, err := vm.EvaluateSnippet(test.name, string(input))
+			if err != nil {
+				t.Fail()
+				t.Errorf("evaluate snippet: %v", err)
+			}
+			if *update {
+				err := ioutil.WriteFile(test.golden, []byte(output), 0666)
+				if err != nil {
+					t.Errorf("error updating golden files: %v", err)
+				}
+				return
+			}
+			golden := read(test.golden)
+			if bytes.Compare(golden, []byte(output)) != 0 {
+				t.Fail()
+				t.Errorf("%s.input != %s\n", test.name, test.golden)
+				data, err := diff(golden, []byte(output))
+				if err != nil {
+					t.Errorf("computing diff: %s", err)
+				}
+				t.Errorf("diff %s jsonnet %s.input\n", test.golden, test.name)
+				t.Errorf(string(data))
+
+			}
+		})
+	}
+}
+
+// "barrowed" from the std library
```

Comment:
@sparkprime how would you feel about adding a library that does the diffing? 
since it's in the tests, it's not a dependency and it would look nicer and would resolve the license issue

**diff**:
![image](https://user-images.githubusercontent.com/429977/29432016-4add916c-839a-11e7-91c7-3a02d791373d.png)

**deps**:
<details>

```
sevki@sevki-ThinkPad-X220:~/code/go/src/github.com/google/go-jsonnet$ go list -f '{{ join .Imports "\n" }}'
bytes
encoding/hex
fmt
math
reflect
sort
strconv
strings
unicode/utf8
sevki@sevki-ThinkPad-X220:~/code/go/src/github.com/google/go-jsonnet$ go list -f '{{ join .Deps "\n" }}'
bytes
encoding/hex
errors
fmt
internal/cpu
internal/poll
internal/race
io
math
os
reflect
runtime
runtime/internal/atomic
runtime/internal/sys
sort
strconv
strings
sync
sync/atomic
syscall
time
unicode
unicode/utf8
unsafe
```
</details>